Sources/Fuzzilli/Corpus/MarkovCorpus.swift

```diff
@@ -0,0 +1,206 @@
+/// Simply put, the corpus keeps track of which paths have been found, and prioritizes seeds
+/// whose path has been hit less than average. Ideally, this allows the fuzzer to prioritize
+/// less explored coverage.
+
+public class MarkovCorpus: ComponentBase, Corpus {
+
+    // All programs currently in the corpus
+    private var allIncludedPrograms: [Program] = []
+    // Queue of programs to be executed next, all of which hit a rare edge
+    private var programExecutionQueue: [Program] = []
+
+    // For each edge encountered thus far, track which program initially discovered it
+    private var edgeMap: [UInt32:Program] = [:]
+
+    // This scheduler tracks the total number of samples it has returned
+    // This allows it to build an initial baseline by randomly selecting a program to mutate
+    // before switching to the more computationally expensive selection of programs that
+    // hit infreqent edges
+    private var totalExecs: UInt32 = 0
+
+    // This scheduler returns one base program multiple times, in order to compensate the overhead caused by tracking
+    // edge counts
+    private var currentProg: Program
+    private var remainingEnergy: UInt32 = 0
+
+    // Markov corpus requires an evaluator that tracks edge coverage
+    // Thus, the corpus object keeps a reference to the evaluator, in order to only downcast once
+    private var covEvaluator: ProgramCoverageEvaluator
+
+    // Rate at which selected samples will be included, to promote diversity between instances
+    // Equivalent to 1 - dropoutRate
+    private var dropoutRate: Double
+
+    // The scheduler will initially selectd the 1 / desiredSelectionProportion samples with the least frequent
+    // edge hits in each round, before dropout is applied
+    private let desiredSelectionProportion = 8
+
+    public init(covEvaluator: ProgramCoverageEvaluator, dropoutRate: Double) {
+        self.dropoutRate = dropoutRate
+        covEvaluator.enableEdgeTracking()
+        self.covEvaluator = covEvaluator
+        self.currentProg = Program()
+        super.init(name: "MarkovCorpus")
+    }
+
+    override func initialize() {
+        assert(covEvaluator === fuzzer.evaluator as! ProgramCoverageEvaluator)
+        currentProg = makeSeedProgram()
+        add(currentProg)
+    }
+
+    // This should only be called when ProgramAspects are not available, such as on an import from
+    // another fuzzer instance
+    public func add(_ program: Program) {
+        guard program.size > 0 else { return }
+        if let aspects = execAndEvalProg(program) {
+            add(program, aspects)
+        }
+    }
+
+    public func add(_ program: Program, _ aspects: ProgramAspects) {
+        guard program.size > 0 else { return }
+
+        guard let origCov = aspects as? CovEdgeSet else {
+            logger.fatal("Markov Corpus needs to be provided a CovEdgeSet when adding a program")
+        }
+
+        covEvaluator.clearEdgeList(origCov.toEdges())
```

Comment:
WDYT about making this `covEvaluator.resetAspects(aspects)` since it's probably not really tied to coverage?

Reply:
Makes sense. It's in the latest commit